crypto/src/mls/buffer_external_commit.rs

```diff
@@ -0,0 +1,191 @@
+use core_crypto_keystore::entities::{EntityFindParams, MlsPendingMessage, PersistedMlsPendingGroup};
+use openmls::prelude::{MlsMessageIn, MlsMessageInBody};
+use tls_codec::Deserialize;
+
+impl MlsCentral {
+    pub(crate) async fn handle_when_group_is_pending(
+        &mut self,
+        id: &ConversationId,
+        message: impl AsRef<[u8]>,
+    ) -> CryptoResult<MlsConversationDecryptMessage> {
+        let keystore = self.mls_backend.borrow_keystore();
+        let Ok(Some(pending_group)) = keystore.find::<PersistedMlsPendingGroup>(id).await else {
+            return Err(CryptoError::ConversationNotFound(id.clone()));
+        };
+
+        let pending_msg = MlsPendingMessage {
+            id: pending_group.id.clone(),
+            message: message.as_ref().to_vec(),
+        };
+        keystore.save::<MlsPendingMessage>(pending_msg).await?;
+        Err(CryptoError::UnmergedPendingGroup)
+    }
+
+    pub(crate) async fn restore_pending_messages(
+        &mut self,
+        conversation: &mut MlsConversation,
+    ) -> CryptoResult<Option<Vec<MlsConversationDecryptMessage>>> {
+        let keystore = self.mls_backend.borrow_keystore();
+
+        let mut pending_messages = keystore
+            .find_all::<MlsPendingMessage>(EntityFindParams::default())
+            .await?
+            .into_iter()
+            .try_fold(vec![], |mut acc, m| {
+                let msg = MlsMessageIn::tls_deserialize_bytes(m.message.as_slice()).map_err(MlsError::from)?;
+                let ct = match msg.body_as_ref() {
+                    MlsMessageInBody::PublicMessage(m) => Ok(m.content_type()),
+                    MlsMessageInBody::PrivateMessage(m) => Ok(m.content_type()),
+                    _ => Err(CryptoError::ImplementationError),
+                }?;
+                acc.push((ct as u8, msg));
+                CryptoResult::Ok(acc)
+            })?;
+
+        // We want to restore application messages first, then Proposals & finally Commits
+        // luckily for us that's the exact same order as the [ContentType] enum
+        pending_messages.sort_by(|(a, _), (b, _)| a.cmp(b));
+
+        let mut decrypted_messages = vec![];
+        for (_, m) in pending_messages {
+            let parent_conversation = if let Some(parent_id) = &conversation.parent_id {
+                Some(
+                    self.get_conversation(parent_id)
+                        .await
+                        .map_err(|_| CryptoError::ParentGroupNotFound)?,
+                )
+            } else {
+                None
+            };
+            let callbacks = self.callbacks.as_ref().map(|boxed| boxed.as_ref());
+            let decrypted = conversation
+                .decrypt_message(m, parent_conversation, self.mls_client()?, &self.mls_backend, callbacks)
+                .await?;
+            decrypted_messages.push(decrypted);
+        }
+
+        let decrypted_messages = (!decrypted_messages.is_empty()).then_some(decrypted_messages);
```

Comment:
Shouldn't we delete the messages that were restored as well? It would make sense to avoid running into storage issues eventually.

Reply:
Done